probe-rs/src/flashing/builder.rs

```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
-                    add_fill(
-                        page_address + fill_start as u32,
-                        fill_size as u32,
-                        &mut fills,
-                        pages.len() - 1,
-                    );
-                }
+            // Ignore if no data.
+            if !sector_has_data && !page_has_data {
+                continue;
+            }
 
-                // Denotes whether a new sector will be done next iteration round.
-                let start_new_sector =
-                    current_block_address + size as u32 >= sector_address + sector_size;
-                // Denotes whether we are done with the flash building process now.
-                let last_bit_of_block = block_offset + size == block.size() as usize
-                    && !self.data_blocks.is_empty()
-                    && n == self.data_blocks.len() - 1;
-
-                // If one of the two conditions resolves to true, and we are including
-                // pages which will only contain fill, then we fill all remaining pages
-                // for the current sector.
-                if (start_new_sector || last_bit_of_block) && include_empty_pages {
-                    // Iterate all possible sector pages and see if they have been created yet.
-                    let pages_per_sector =
-                        (sector_size / flash_algorithm.flash_properties.page_size) as usize;
-                    'o: for i in 0..pages_per_sector {
-                        // Calculate the possible page address.
-                        let page_address =
-                            sector_address + i as u32 * flash_algorithm.flash_properties.page_size;
-                        // Get the maximum available already added pages up to a maximum of
-                        // the available pages per sector.
-                        let last_pages_num_max = pages_per_sector.min(pages.len());
-                        // Get those pages from the pages vector.
-                        let max_last_pages = &pages[pages.len() - last_pages_num_max..];
-                        for page in max_last_pages {
-                            if page.address == page_address {
-                                continue 'o;
-                            }
-                        }
-                        let page = add_page(flash_algorithm, page_address, &mut pages)?;
-                        add_fill(page.address, page.size(), &mut fills, pages.len() - 1);
-                    }
+            sectors.push(FlashSector {
+                address: info.base_address,
+                size: info.size,
+            })
+        }
+
+        for info in flash_algorithm.iter_pages() {
+            let page_end = info.base_address + info.size;
+            let range = info.base_address..page_end;
+
+            // Ignore if outside the NvmRegion.
+            if !region.range.contains_range(&range.clone()) {
```

Comment:
How can a page ever not be contained in a memory range?

Reply:
We're iterating all the pages covered by the flash algorithm, which may be more than the NvmRegion if the algo manages multiple NvmRegions.

Reply:
Ah, I see! Thanks!